File: js-template/leetcode/editor/cn/3-sum.js
/*
 * @lc app=leetcode.cn id=15 lang=javascript
 * @lcpr version=30403
 *
 * [15] 三数之和
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    const result = [];
    const sorted = nums.sort((a, b) => a - b);

    for(let i = 0; i <  sorted.length; i++) {
        if (sorted[i] > 0) break;
        if (i > 0 && sorted[i] === sorted[i - 1]) continue;
        let low = i + 1;
        let high = sorted.length - 1;
        if (sorted[high] < 0) break;
        while(low < high) {
            if (sorted[low] === sorted[low-1] && low > i+1)
                low++;
            else if (high < sorted.length - 1 && sorted[high] === sorted[high+1]) {
                high--;
            } else {
                const sum = sorted[i] + sorted[low] + sorted[high];
                if (sum === 0) {
                    result.push([sorted[i], sorted[low], sorted[high]]);
                    high--;
                    low++;
                } else if (sum > 0) {
                    high--;
                } else if (sum < 0) {
                    low++;
                }
            }
        }
    }
    
    return result;
};

// @lc code=end

// your test code here





/*
// @lcpr case=start
// [-1,0,1,2,-1,-4]\n
// @lcpr case=end

// @lcpr case=start
// [0,1,1]\n
// @lcpr case=end

// @lcpr case=start
// [0,0,0]\n
// @lcpr case=end

// @lcpr case=start
// [0,0,0,0]\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/3-sum.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 * @lcpr version=30404
 *
 * [15] 三数之和
 */

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <optional>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        optional<int> pre;
        
        sort(nums.begin(), nums.end());

        for (auto i = 0; i < nums.size() - 2; ++i) {
            if (pre.has_value() && pre.value() == nums[i])
                continue;
            if (nums[i] > 0)
                break;
            int left = i + 1, right = nums.size() - 1;
            while (left < right)
            {
                int lValue = nums[left];
                int rValue = nums[right];
                int sum = nums[i] + lValue + rValue; 
                if (0 == sum) {
                    result.push_back({nums[i], lValue, rValue});
                    while (lValue == nums[left] && left < right)
                    {
                        ++left;
                    }
                    while (rValue == nums[right] && left < right)
                    {
                        --right;
                    }
                } else if (0 > sum) {
                    while (lValue == nums[left] && left < right)
                    {
                        ++left;
                    }
                } else if (0 < sum) {
                    while (rValue == nums[right] && left < right)
                    {
                        --right;
                    }
                }
            }
            
            pre = nums[i];
        }

        return result;
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    // vector<int> nums = {-1,0,1,2,-1,-4};
    vector<int> nums = {0,0,0,0, 0,0};
    solution.threeSum(nums);
}



/*
// @lcpr case=start
// [-1,0,1,2,-1,-4]\n
// @lcpr case=end

// @lcpr case=start
// [0,1,1]\n
// @lcpr case=end

// @lcpr case=start
// [0,0,0]\n
// @lcpr case=end

 */

